### ISSUE TITLE
Canvas rendering artifacts when dragging brush mark

### ISSUE THREAD

**Issue by arvind** [vega/vega#813]
![canvasrendering](https://cloud.githubusercontent.com/assets/42262/26123511/fe5459c4-3a2f-11e7-8c2e-a3d182f5be88.gif)

With the canvas renderer, Vega 3.0.0-beta.31, and the following spec (apologies for the verbosity, but tried to make it as minimal as possible): 

```json
{
  "$schema": "http://vega.github.io/schema/vega/v3.0.json",
  "description": "A scatterplot showing horsepower and miles per gallons for various cars.",
  "autosize": "pad",
  "padding": 5,
  "data": [
    {"name": "foo_store"},
    {
      "name": "source_0",
      "url": "data/cars.json",
      "format": {
        "type": "json",
        "parse": {
          "Horsepower": "number",
          "Miles_per_Gallon": "number"
        }
      },
      "transform": [
        {
          "type": "filter",
          "expr": "datum[\"Horsepower\"] !== null && !isNaN(datum[\"Horsepower\"]) && datum[\"Miles_per_Gallon\"] !== null && !isNaN(datum[\"Miles_per_Gallon\"])"
        }
      ]
    }
  ],
  "signals": [
    {"name": "width","update": "200"},
    {"name": "height","update": "200"},
    {
      "name": "unit",
      "value": {},
      "on": [
        {
          "events": "mousemove",
          "update": "group()._id ? group() : unit"
        }
      ]
    }
  ],
  "marks": [
    {
      "name": "nested_main_group",
      "type": "group",
      "signals": [
        {
          "name": "foo_Horsepower",
          "value": [],
          "on": [
            {
              "events": {
                "source": "scope",
                "type": "mousedown",
                "filter": [
                  "!event.item || (event.item && event.item.mark.name !== \"foo_brush\")"
                ]
              },
              "update": "invert(\"x\", [x(unit), x(unit)])"
            },
            {
              "events": {
                "source": "window",
                "type": "mousemove",
                "consume": true,
                "between": [
                  {
                    "source": "scope",
                    "type": "mousedown",
                    "filter": [
                      "!event.item || (event.item && event.item.mark.name !== \"foo_brush\")"
                    ]
                  },
                  {"source": "window","type": "mouseup"}
                ]
              },
              "update": "[foo_Horsepower[0], invert(\"x\", clamp(x(unit), 0, width))]"
            },
            {
              "events": {"signal": "foo_translate_delta"},
              "update": "clampRange([foo_translate_anchor.extent_x[0] + abs(span(foo_translate_anchor.extent_x)) * foo_translate_delta.x / foo_translate_anchor.width, foo_translate_anchor.extent_x[1] + abs(span(foo_translate_anchor.extent_x)) * foo_translate_delta.x / foo_translate_anchor.width], invert(\"x\", 0), invert(\"x\", unit.width))"
            }
          ]
        },
        {
          "name": "foo_size",
          "value": [],
          "on": [
            {
              "events": {
                "source": "scope",
                "type": "mousedown",
                "filter": [
                  "!event.item || (event.item && event.item.mark.name !== \"foo_brush\")"
                ]
              },
              "update": "{x: x(unit), y: y(unit), width: 0, height: 0}"
            },
            {
              "events": {
                "source": "window",
                "type": "mousemove",
                "consume": true,
                "between": [
                  {
                    "source": "scope",
                    "type": "mousedown",
                    "filter": [
                      "!event.item || (event.item && event.item.mark.name !== \"foo_brush\")"
                    ]
                  },
                  {"source": "window","type": "mouseup"}
                ]
              },
              "update": "{x: foo_size.x, y: foo_size.y, width: abs(x(unit) - foo_size.x), height: abs(y(unit) - foo_size.y)}"
            }
          ]
        },
        {
          "name": "foo",
          "update": "[{encoding: \"x\", field: \"Horsepower\", extent: foo_Horsepower}]"
        },
        {
          "name": "foo_translate_anchor",
          "value": {},
          "on": [
            {
              "events": [
                {
                  "source": "scope",
                  "type": "mousedown",
                  "markname": "foo_brush"
                }
              ],
              "update": "{x: x(unit), y: y(unit), width: foo_size.width, height: foo_size.height, extent_x: slice(foo_Horsepower), }"
            }
          ]
        },
        {
          "name": "foo_translate_delta",
          "value": {},
          "on": [
            {
              "events": [
                {
                  "source": "window",
                  "type": "mousemove",
                  "consume": true,
                  "between": [
                    {
                      "source": "scope",
                      "type": "mousedown",
                      "markname": "foo_brush"
                    },
                    {"source": "window","type": "mouseup"}
                  ]
                }
              ],
              "update": "{x: x(unit) - foo_translate_anchor.x, y: y(unit) - foo_translate_anchor.y}"
            }
          ]
        },
        {
          "name": "foo_tuple",
          "on": [
            {
              "events": {"signal": "foo"},
              "update": "{unit: unit.datum && unit.datum._id, intervals: foo}"
            }
          ]
        },
        {
          "name": "foo_modify",
          "on": [
            {
              "events": {"signal": "foo"},
              "update": "modify(\"foo_store\", foo_tuple, true)"
            }
          ]
        }
      ],
      "encode": {
        "update": {
          "width": {"signal": "width"},
          "height": {"signal": "height"},
          "fill": {"value": "transparent"}
        }
      },
      "marks": [
        {
          "name": "foo_brush",
          "type": "rect",
          "encode": {
            "enter": {
              "fill": {"value": "black"},
              "stroke": {"value": "white"}
            },
            "update": {
              "x": [
                {
                  "test": "data(\"foo_store\").length && foo_tuple && foo_tuple.unit === data(\"foo_store\")[0].unit",
                  "scale": "x",
                  "signal": "foo[0].extent[0]"
                },
                {"value": 0}
              ],
              "x2": [
                {
                  "test": "data(\"foo_store\").length && foo_tuple && foo_tuple.unit === data(\"foo_store\")[0].unit",
                  "scale": "x",
                  "signal": "foo[0].extent[1]"
                },
                {"value": 0}
              ],
              "y": [
                {
                  "test": "data(\"foo_store\").length && foo_tuple && foo_tuple.unit === data(\"foo_store\")[0].unit",
                  "value": 0
                },
                {"value": 0}
              ],
              "y2": [
                {
                  "test": "data(\"foo_store\").length && foo_tuple && foo_tuple.unit === data(\"foo_store\")[0].unit",
                  "field": {"group": "height"}
                },
                {"value": 0}
              ]
            }
          }
        }
      ]
    }
  ],
  "scales": [
    {
      "name": "x",
      "type": "linear",
      "domain": {"data": "source_0","field": "Horsepower"},
      "range": [0,200],
      "round": true,
      "nice": true,
      "zero": true
    },
    {
      "name": "y",
      "type": "linear",
      "domain": {"data": "source_0","field": "Miles_per_Gallon"},
      "range": [200,0],
      "round": true,
      "nice": true,
      "zero": true
    }
  ]
}
```

**Comment by jheer** [comment]
So, it turns out this is not a simple bounds calculation issue. This is a subtle bug that arises due to an interaction that spans multiple modules.

_What is happening:_ the brush is being "updated" on two subsequent pulses prior to rendering. The first pulse occurs due to mouse input. The second pulse occurs due to a `runAfter` call made by the `modify` expression, which is then processed _prior_ to rendering. On the second pulse, the brush is not actually changing, but the current encoder implementation treats any mark with an invoked encoder as "modified", even if the encodings do not change the value of any item properties. The "modified" brush is then passed through bounds calculation a second time, overwriting the correct previous bounds. When the item is finally handed over for rendering, the incorrect bounds information leads to insufficient clipping, causing the observed artifacts.

This leaves us with some different potential follow-ups:
1. Adjust vega-parser to emit encoder logic that checks prior values. Only include an item in the pulse.mod array if the encoder changes one or more values. This should prevent such items from propagating forward for an erroneous secondary bounds calculation.
2. Revisit how rendering vs. runAfter is handled. Right now, a reinvocation of the dataflow can be triggered within a runAfter callback, which will be processed prior to rendering. This is more performant in that we don't "needlessly" render multiple times on a single interaction loop, but it does make the renderer vulnerable to corrupted bounds state. Not sure what changes we can or should make here, but some change may be possible.
3. Change how we handle previous bounds. For example, instead of clearing the prior bounds we could aggregate the bounds between render calls. However, this would require additional bookkeeping and/or dependencies to know on what pulse stamp the renderer was last invoked.

I'll think about this more, but at least this pins down the issue.

**Comment by jheer** [comment]
I think I found a clean solution that also reduces Vega's memory footprint: remove tracking of previous bounds altogether. The `Bound` operator can simply report previous bounds to the renderer before computing new bounds. This allows the renderer to always know the damaged region, and scenegraph items need not carry around their prior state. Now we just need to make sure something else doesn't get broken in the process...

**Comment by jheer** [comment]
Fixed in vega-scenegraph v2.0.0-beta.29 and vega-view v1.0.0-beta.32 via overhaul of previous bounds and dirty item handling.
